WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java

```diff
@@ -0,0 +1,121 @@
+        final byte[] key = new byte[XCHACHA20POLY1305_KEYBYTES];
+        NaCl.sodium().crypto_secretstream_xchacha20poly1305_keygen(key);
+
+        final String encryptedKeyBase64 = getBoxSealEncryptedBase64String(
+                publicKeyBase64,
+                key,
+                XCHACHA20POLY1305_KEYBYTES);
+        encryptionDataJson.put("encryptedKey", encryptedKeyBase64);
+
+        final byte[] state = new byte[XCHACHA20POLY1305_STATEBYTES];
+
+        final String headerBase64 = initSecretStreamXchacha20poly1305(state, key);
+        encryptionDataJson.put("header", headerBase64);
+
+        JSONArray encryptedElementsJson = new JSONArray();
+        if (!stringData.isEmpty()) {
+            final String[] splitStringData = stringData.split("\n"); // break up the data by line
+            
+            for (int i = 0; i < splitStringData.length; ++i) {
+                String element = splitStringData[i];
+                element = element + "\n";
+                final String encryptedElementBase64 = getSecretStreamXchacha20poly1305EncryptedBase64String(
+                        state,
+                        element,
+                        XCHACHA20POLY1305_TAG_MESSAGE);
+                encryptedElementsJson.put(encryptedElementBase64);
+            }
+        }
+
+        final String encryptedDataBase64 = getSecretStreamXchacha20poly1305EncryptedBase64String(
+                state,
+                "",
+                XCHACHA20POLY1305_TAG_FINAL);
+        encryptedElementsJson.put(encryptedDataBase64);
+
+        encryptionDataJson.put("messages", encryptedElementsJson);
+
+        return encryptionDataJson.toString();
+    }
+
+    private static String getBoxSealEncryptedBase64String(final String publicKeyBase64,
+                                                          final byte[] data,
+                                                          final int dataSize) {
+        final byte[] encryptedData = new byte[dataSize + BOX_SEALBYTES];
+        final byte[] publicKeyBytes = Base64.decode(publicKeyBase64, Base64.DEFAULT);
+        NaCl.sodium().crypto_box_seal(encryptedData, data, dataSize, publicKeyBytes);
+        return Base64.encodeToString(encryptedData, BASE64_ENCODE_FLAGS);
+    }
+
+    private static String initSecretStreamXchacha20poly1305(byte[] state, final byte[] key) {
+        final byte[] header = new byte[XCHACHA20POLY1305_HEADERBYTES];
+        NaCl.sodium().crypto_secretstream_xchacha20poly1305_init_push(state, header, key);
+
+        return Base64.encodeToString(header, BASE64_ENCODE_FLAGS);
+    }
+
+    private static String getSecretStreamXchacha20poly1305EncryptedBase64String(final byte[] state,
+                                                                                final String data,
+                                                                                final short tag) {
+        final int[] clen = new int[0];
+        final byte[] ad = new byte[0];
```

Comment:
These variable names are a little unclear – could you give them more descriptive names?

Reply:
yep!  Thank you for catching that. 🙂  Fixed this in: a57da8fcf15c4634b8d2811c3f79275bc6d2130a